File: SdkGenerator/Model/Example/SettingExample/SettingValue.cs
namespace SdkGenerator.Model.Example.SettingExample;

public class SettingValue<T>
{
    public bool Inherited { get; set; }
    public required T Value { get; set; }
}

public class ToggleableSetting<T> : SettingValue<T>
{
    public bool Enabled { get; set; }
}

public class NullableSettingValue<T>
{
    public bool Inherited { get; set; }
    public T? Value { get; set; }
    public bool HasValue() => Value is not null;
}

public class NullableToggleableSetting<T> : NullableSettingValue<T>
{
    public bool Enabled { get; set; }

    public T? ComputedValue() => Enabled ? Value : default;
}

File: Templates/TypeScript/Settings/setting-value.ts
import { JsonProperty } from "json-object-mapper";

export class SettingValue<T> {
	@JsonProperty({ name: "inherited" })
	public inherited: boolean;

	@JsonProperty({ required: true })
	public value: T;
}

export class ToggleableSetting<T> extends SettingValue<T> {
	@JsonProperty({ required: true })
	public enabled: boolean;
}

export class NullableSettingValue<T> {
	@JsonProperty({ name: "inherited" })
	public inherited: boolean;

	@JsonProperty({ required: false })
	public value?: T;

	public hasValue(): boolean {
		return this.value !== undefined;
	}
}

export class NullableToggleableSetting<T> extends NullableSettingValue<T> {
	@JsonProperty({ required: true })
	public enabled: boolean;

	public computedValue(): T | undefined {
		return this.enabled ? this.value : undefined;
	}
	//TODO: Figure out default value
	//public T? ComputedValue() => Enabled ? Value : default;
}
